Add Python extension using pocketpy

hcarty opened the pull request:
This adds Python bindings to orx through an extension, using https://pocketpy.dev/ to provide a fully embedded Python runtime.

To quote the pocketpy website:

> pocketpy is a portable Python 3.x interpreter, written in C11. It aims to be an alternative to Lua for game scripting, with elegant syntax, powerful features and competitive performance. pocketpy has no dependencies other than the C standard library, which can be easily integrated into your C/C++ project. Developers are able to write Python bindings via C-API

### Extension features
- Adds a `+python` option to orx's project init tool
    - This will initialize a new project so that it can run fully through Python, including `init`, `update`, and `exit` engine callbacks and several event handlers.
- Event handlers
    - Object creation, deletion
    - Physics collision, separation
    - Setting shader parameter values
- Bindings to several engine functions under an `orx.*` module namespace, including:
    - `orxObject_*` in `orx.object`
    - `orxConfig_*` in `orx.config`
    - `orxInput_*` in `orx.input`
    - `orxVector_*` in `orx.vector`
    - `orxClock_ComputeDT` in `orx.clock`
    - `orxCommand_Evaluate(WithGUID)` in `orx.command`
    - `orxLOG` as `orx.log`
- `*.pyi` files for each of the provided Python modules to improve support for editor integration like VSCode's Python extension and support Python type checkers like pyright and mypy
- Registers `Python.Exec` command to run Python code from commands
- A Scroll-inspired `pyscroll` module to bind per-object behaviors
- Uses `orxResource` module for loading `*.py` sources. Python modules can be included from any supported resource location, including bundles when the `bundle` extension is used.
- Works for native and web/emscripten builds
- The full pocketpy runtime and standard library is compiled directly into the application binary with no additional runtime requirements
- Custom function and data bindings can be added using pocketpy's C API

hcarty committed:
Add python extension

- Based on pocketpy (https://pocketpy.dev/)

hcarty commented on code/build/template/[+python pyproject.toml]:
```diff
@@ -0,0 +1,2 @@
+[tool.pyright]
```
This is used by `pyright`, a type checker for Python. There are no `*.py` sources for the core engine bindings so the type checker will warn or error on some settings otherwise.

iarwain replied:
Can't this be initialized programmatically instead?

hcarty replied:
I can remove the configuration file. It's something that was more necessary when I started the extension, but the tooling has advanced since then and people may want to use other tooling for their own projects.

iarwain replied:
By that I meant, could this be expressed in an orx config file instead, and set programmatically when starting the extension?
This way we keep all configuration stuff consistent across engine & extensions.

hcarty commented on code/build/template/build/premake4.lua:
```diff
@@ -107,6 +111,12 @@ solution "[name]"
+
+    defines
+    {
+        "PK_ENABLE_OS=0"
```
This leaves out some system interaction modules and functions from pocketpy. This could be changed by a user locally, but leaving IO to orx seems like a good default.

iarwain replied:
Can this be moved to orxPy.h instead? I'd rather keep things as self contained as possible without relying on external defines.

hcarty commented on code/build/template/build/premake4.lua:
```diff
@@ -97,6 +97,10 @@ solution "[name]"
     {
+[+python
+        "C11",]
+[+imgui +sndh
```
This replaces the C++ version and linking flags later in the file for these extensions.

hcarty commented on code/build/template/build/premake4.lua:
```diff
@@ -284,6 +290,8 @@ solution "[name]"
         buildoptions
         {
+[+python
+            "/utf-8",]
```
pocketpy's docs specify that this flag is needed for MSVC

hcarty commented on code/build/template/data/[+python python]/[=name].py:
This file defines and associates the main Python callbacks with the names registered in the C portion of the extension.

hcarty commented on code/build/template/data/[+python python]/miniscroll.py:
Heavily Scroll-inspired!

iarwain replied:
How about `pyscroll` instead?

hcarty commented on code/build/template/data/[+python python]/orx/__init__.pyi:
All of the `*.pyi` files define the interface/type signature of the modules defined in `orxPy.h`. They are here for reference and for use from Python tooling that supports them.

hcarty commented on code/build/template/data/[+python python]/orx/guid.pyi:
```diff
@@ -0,0 +1,4 @@
+class Guid:
```
This is a custom wrapper type since Python does not have a concept of unsigned types, and pocketpy uses `int64_t` internally for Python `int` values.

hcarty commented on code/build/template/include/extensions/[+python orxPy.h]:
The custom bindings between orx and pocketpy are all defined in this file.

hcarty commented on code/build/template/include/extensions/[+python pocketpy]/pocketpy.h:
This is pocketpy's header. The only modification from upstream is the reference to orx's memory allocation functions.

hcarty commented on code/build/template/src/[+python pocketpy]/pocketpy.c:
This is pocketpy's implementation. It is taken directly from upstream without modification.

iarwain replied:
This should be moved inside `include/extensions`.

hcarty replied:
I started with that but it breaks compilation of C++ projects since the code does not compile as C++.

iarwain replied:
Ah I see, in that case let's move it inside the newly added `extensions` subdirectory.

hcarty replied:
Very nice change - thank you! I moved `src/pocketpy/` into `src/extensions/pocketpy/`.

hcarty commented on code/build/template/src/[=name].c[+c++ pp]:
I've successfully tested this locally with the following combinations:
- `+python`
- `+python -scroll`
- `+python -scroll -c++`

hcarty commented on code/build/template/include/extensions/orxExtensions.h:
```diff
@@ -91,6 +96,10 @@ void InitExtensions()
 ]
+[+python
+  // Initialize Python support
+  orxPy_InitVM();
```
Initialize the pocketpy virtual machine. This does not execute any game-specific Python code. It does include all of the `orx.*` engine module bindings.

hcarty commented on code/build/template/src/[=name].c[+c++ pp]:
```diff
@@ -110,18 +122,30 @@ __declspec(dllexport) int AmdPowerXpressRequestHighPerformance = 1;
+[+python
+
+  // Call Python init function
+  eStatus = orxPy_Init();]
```
This calls the user's/game's `orx.init` function, called from the init project's main source file.

iarwain replied:
Shouldn't this call also be inside `orxExtensions.h`, in `InitExtensions()`?
Is there a reason why it is separate from `orxPy_InitVM()`?

hcarty replied:
I started with this combined as well, but if a project defines its own modules from C then those need to be defined before and game code is run. Splitting the two initializes the VM early, then gives space for custom C bindings and module definitions.

iarwain replied:
Does the VM rely on any orx API? If not, we could move its initialization to the BootStrap function instead.
I'd rather not have any extension code outside of orxExtensions.h as much as possible as it breaks adding/removing extensions (the main source file can't be touched once created).

hcarty replied:
Short version of all the text below: I think we can make this change. I have some minor concerns around more advanced use cases described below. If it's ok for me to change the `InitExtensions` return type then I think my main concerns would be addressed.

The call to `orxPy_Init` here is like the call to `orxPy_Update` and `orxPy_CameraUpdate` earlier in the file. They each call Python functions defined by game code at the engine-appropriate time. I've tested locally and `orxPy_Init` could be called from `InitExtensions` if `orxPy_InitVM` is called from `BootstrapExtensions`.

I have some concerns with that change:
- `InitExtensions` doesn't return a status value. Would it be ok to have `InitExtensions` return `orxSTATUS` rather than `void`? In the PR's current state a failure during `orxInit_Py` will stop the game from starting up. I think that's a good feature to have but I'm not sure if others would agree.
- It obfuscates where the game's Python code is loaded and the user's `orx.init` callback function is called. I think that's ok. The Python extension is kind of like Scroll in that it can be used to change where primary game logic is run (C vs C++ vs C++/Scroll vs Python). However the Python extension uses the existing C/C++/Scroll engine callbacks for bootstrap/init/update/camera update rather than defining its own.
- It is less clear where a user's custom C to Python bindings should be added. If `orxPy_Init` is called from `InitExtensions` then those modules must be defined before the call to `InitExtensions`. That should work as long as those module definitions do not rely on other extensions being initialized first. Custom module definitions shouldn't require that - initialization should happen when the game logic is initialized. I think that's ok as well.

A bit of extra context on the extension's two init functions:

`orxPy_InitVM` initializes the pocketpy VM and registers the `orx.*` Python modules with that VM. It does not run any Python code or call any orx APIs. It could be called from `BootstrapExtensions`.

`orxPy_Init` loads and runs the user's `[=name].py` code, sets up any references or bindings between user code and the pocketpy VM, and calls the user-defined Python `init` function if one has been defined. Arbitrary orx and Python code can be run at this point.

iarwain replied:
I'll have to check `orxPy.h`, but shouldn't it no rely on Scroll at all and be orthogonal to it, so as to have a single code flow no matter what the kind of project is?

hcarty reviewed (commented):
Adding some review comments.

iarwain commented on code/build/template/build/premake4.lua:
```diff
@@ -132,11 +142,13 @@ solution "[name]"
         links {"orxp"}
 
     configuration {"*Release*"}
+[+python
```
Same for `NDEBUG`, it should be handled from within `orxPy.h` (cf. orxSoundSystem.c) and not leak from there as well.

hcarty replied:
For `PK_ENABLE_OS` and `NDEBUG`, these affect `pocketpy.c` which doesn't include `orxPy.h`. I could edit `pocketpy.h` with these changes, and adapt that file on new pocketpy version updates. For this PR I've tried to leave the upstream `pocketpy.h` and `pocketpy.c` files as untouched as possible.

Would you prefer having a change in `pocketpy.h` to derive these defines from orx's debug state? That would require some additional adaptation for each new version of pocketpy, but would also eliminate the global defines from `premake4.lua`.

iarwain commented on code/build/template/include/extensions/[+python pocketpy]/pocketpy.h:
```diff
@@ -0,0 +1,1029 @@
+
+// clang-format off
+
+#define PK_VERSION				"2.1.0"
```
Minor alignment issue. Is this file auto-generated?

hcarty replied:
This file comes from the pocketpy project. pocketpy generates the combined `pocketpy.h` and `pocketpy.c` files from individual header and source files as part of the release process.

hcarty reviewed (commented):
Added a few more small review comments.

iarwain commented on code/build/template/build/premake4.lua:
```diff
@@ -132,11 +145,13 @@ solution "[name]"
     configuration {"*Profile*"}
         targetsuffix ("p")
-        defines {"__orxPROFILER__"}
+        defines {"__orxPROFILER__"[+python  ,"NDEBUG"]}
```
You can add a separate `defines` line for Python, it might make it slightly more easy to read.
Or just decompose the existing one on multiple lines.

iarwain commented on code/build/template/data/[+python python]/[=name].py:
```diff
@@ -0,0 +1,53 @@
+orx.on_camera_update = camera_update
+
+# Setup engine callbacks for miniscroll
+miniscroll.setup(init, update)
```
Should we give it a more obvious name maybe, like PyScroll?

hcarty replied:
I like PyScroll! I started with "mini" as the prefix because of its relatively limited scope, but I think PyScroll is better and disambiguates well with Scroll.

iarwain commented:
> * A Scroll-inspired `miniscroll` module to bind per-object behaviors

Wouldn't `pyscroll` be a better suited name in that case?

> * The full pocketpy runtime and standard library is compiled directly into the application binary with no additional runtime requirements

Love that part! :)

Additionally, would there be any benefit to expose orx's new task worker pool to the Python environment?

iarwain commented on code/build/template/data/config/[=name].ini:
```diff
@@ -95,6 +100,11 @@ ChildList       = Subtitle
 [+inspector
 OnCreate        = Inspector.RegisterObject ^]
+[+python
+Input           = @
```
Why not use the TriggerList here as well?

hcarty replied:
miniscroll/pyscroll does not support the same `.` and other prefixes that Scroll does, so it can't match the behavior the scroll provides with the `TriggerList`. It's something I could probably add, but last time I looked there were some features Scroll relies on for this that were not bound by orxpy.

iarwain replied:
Ah I see, so the inputs are handled entirely in Python code then?

Fine by me, but I do find the data-only input handling handy for simple situations and I would recommend implementing something similar to what's inside ScrollBase (it's not much code).

hcarty replied:
> so the inputs are handled entirely in Python code then?

Somewhat - it uses a similar approach to what is done in Scroll, but not as refined. This is run for each orx object registered with pyscroll/miniscroll on each update:

```python
      # Input triggers
      for input_name in mini.input_names:
        if orx.input.has_been_activated(input_name):
          mini.o.fire_trigger("Input", refinement=[input_name])
```

The `mini.input_names` list is populated when the object is created, based on the input set associated with the object, similar to what's done by Scroll but captured on object creation rather than looking up the inputs at runtime.

I agree the additional functionality and flexibility provided by Scroll is worth having here. I'll work on that.

hcarty replied:
> You can add a separate defines line for Python, it might make it slightly more easy to read.
> Or just decompose the existing one on multiple lines.

I decomposed the existing one on multiple lines and am happy to change the approach further if you prefer.

iarwain commented on code/build/template/src/[=name].c[+c++ pp]:
```diff
@@ -61,19 +61,24 @@ __declspec(dllexport) int AmdPowerXpressRequestHighPerformance = 1;
 ]
+[+python
+  // Call Python update
+  [-scroll orxPy_Update(_pstClockInfo, orxNULL);][+scroll orxPy_Update(&_rstClockInfo, orxNULL);]]
```
if the only difference is the first parameter, you can have the -scroll/+scroll only for that parameter instead, for brevity.

iarwain commented on code/build/template/src/[=name].c[+c++ pp]:
```diff
@@ -94,6 +99,10 @@ __declspec(dllexport) int AmdPowerXpressRequestHighPerformance = 1;
+]
+[+python
+  // Call Python camera update
+  [-scroll orxPy_CameraUpdate(_pstClockInfo, orxNULL);][+scroll orxPy_CameraUpdate(&_rstClockInfo, orxNULL);]]
```
Similar suggestion as before: use -scroll/+scroll only around the first parameter.

hcarty replied:
Done for both.

iarwain commented on code/build/template/src/[=name].c[+c++ pp]:
```diff
@@ -104,6 +113,7 @@ __declspec(dllexport) int AmdPowerXpressRequestHighPerformance = 1;
   // Init extensions
   InitExtensions();
 
+[-python
```
How are the viewports created with the Python version?

hcarty replied:
The viewports are created in Python, in `data/[+python python]/[=name].py`.

iarwain replied:
One advantage of having them data-driven is, during debug, the ability to have additional viewports to display "hidden" data, without having to modify your code.

hcarty replied:
The Python implementation follows what would be done in the main C or C++ code so it should still be data driven in the same way:
```python
  # Push [Main] as the default config section
  orx.config.push_section("Main")

  # Create viewports defined in [Main]
  for i in range(0, orx.config.get_list_count("ViewportList")):
    orx.viewport.create_from_config(orx.config.get_string("ViewportList", i))
```

iarwain commented on code/build/template/src/[=name].c[+c++ pp]:
```diff
@@ -112,17 +122,22 @@ __declspec(dllexport) int AmdPowerXpressRequestHighPerformance = 1;
-
+]
 [+scroll
+[-python
```
All the -python blocks could be merged in a single one and the callbacks registered at the very end instead.

hcarty replied:
Done.

iarwain reviewed (commented):
Added some comments. :)

hcarty committed:
Updated python extension:

- PR style feedback

hcarty commented:
> Wouldn't `pyscroll` be a better suited name in that case?

I think that's reasonable. I went back and forth in my own head a lot about the naming for this. pyscroll is probably a better name.

I currently use `Mini` as the name for the `ScrollObject` equivalent in Python. Would you be ok with me using `ScrollObject` as the class name in `pyscroll` as well? Or would you prefer a separate name like `PyScroll` or `PyScrollObject` to avoid ambiguity in discussions about code?

>> * The full pocketpy runtime and standard library is compiled directly into the application binary with no additional runtime requirements
>
> Love that part! :)

I do too! That's the main incentive I had to try this out in the first place. No separate runtime to worry about, or binary compatibility issues.

> Additionally, would there be any benefit to expose orx's new task worker pool to the Python environment?

There may be, but I don't think it's needed at this point.

pocketpy supports multiple VMs running in parallel, with one VM per thread. It uses C11 threading, manages its own threads, and that does work with this extension. None of the orx modules are available from any VM other than the main VM though, to avoid issues with the config system and object creation.

The primitives pocketpy provides for communication and tasking between VMs are very minimal so I eventually want to add some friendly wrappers on top. Those may be modeled after the task worker pool.

hcarty commented:
> > Wouldn't `pyscroll` be a better suited name in that case?
> 
> I think that's reasonable. I went back and forth in my own head a lot about the naming for this. pyscroll is probably a better name.
> 
> I currently use `Mini` as the name for the `ScrollObject` equivalent in Python. Would you be ok with me using `ScrollObject` as the class name in `pyscroll` as well? Or would you prefer a separate name like `PyScroll` or `PyScrollObject` to avoid ambiguity in discussions about code?

Thinking about this more, I don't think `ScrollObject` is a good choice on the Python side since the Scroll and pocketpy extensions can be freely mixed in projects - and my hope is that both will be used side by side in at least some cases.

Maybe something like `Base` or `Bound`? I haven't found a name I'm completely happy with yet.

hcarty committed:
Updated python extension:

- Renamed miniscroll to pyscroll
- Renamed Mini class to Base

hcarty commented:
I renamed `Mini` to `Base` and `miniscroll` to `pyscroll` in [3fd6ff6](https://github.com/orx/orx/pull/124/commits/3fd6ff6f64e1d4fbda1b59391d29e0538024abb8).

hcarty committed:
Updated python extension:

- Add orx.input.has_new_status (orxInput_HasNewStatus)
- orx.object.fire_trigger returns event success/failure status
- pyscroll now more closely matches Scroll's config-based input trigger handling

hcarty commented:
[d2afc53](https://github.com/orx/orx/pull/124/commits/d2afc53dcf0b1c1eebb3958813360801de5f6bad) brings pyscroll's config-based input handling more inline with what Scroll supports, including `.` and `-` prefixes and input value refinements.

hcarty committed:
Updated python extension:

- Fix pyscroll input trigger event handling
- Add orx.input.get_next binding orxInput_GetNext

iarwain reviewed (approved):
Thanks for the changes. Let's merge it post 1.17 release!

hcarty committed:
Updated Python extension:

- De-conflict Scroll and pyscroll input handling by not registering the Logo object with pyscroll in a fresh init project with both +python and +scroll
- Refactor pyscroll to be a bit cleaner and more friendly to common Python idioms
- Added and expanded doc comments for pyscroll
- Update to latest pocketpy

hcarty commented:
Thank you for the review!

I pushed that refactor that I mentioned on discord in https://github.com/orx/orx/pull/124/commits/d72da6ae9070080798031bbdf583cfe9faecf8f8. I'm happy to have this merged after the 1.17 release!

iarwain commented on code/build/template/include/extensions/[+python orxPy.h]:
It's super minor, but given this already follows most of orx's coding conventions, would you mind going the extra mile and prefixing function parameters with `_`?
For example:
`orxCHAR *orxPy_ReadSourceFromMain(const orxSTRING _zPath, int *_piDataSize)`

hcarty replied:
Certainly - apologies for the inconsistency for those function parameter names. I'll fix that and push an updated commit in a few minutes.

hcarty replied:
Style fixes are in https://github.com/orx/orx/pull/124/commits/687f58968d0292569711c0dd8346d949619e1552. There were a lot! Thank you for calling the naming mismatch out. I'd like this to feel like orx as much as possible.

iarwain replied:
Thanks for taking care of it!

hcarty committed:
Updated python extension:

- Fix function parameter names to match orx's coding conventions

hcarty committed:
Updated python extension:

- Bugfix in pyscroll's shader parameter handling

iarwain commented on code/build/template/include/extensions/orxExtensions.h:
```diff
@@ -100,14 +105,26 @@ void InitExtensions()
+[+cheat
+  // Exit from cheat support
+  orxCheat_Exit();
+]
```
I believe this is duplicated now, probably a side effect of me rearranging the sequence on the main branch.
Could you remove it, please?

@hcarty Is there anything else pending or are we good to merge the PR?

hcarty replied:
Ah! I apologize for missing the duplication. Thank you for fixing it!

iarwain committed:
Merge branch 'master' into pocketpy-python-extension

iarwain reviewed (approved):
Forgot to submit the review once again. This workflow really doesn't agree with me for some reason. ^^

I fixed the merge issue in the meantime. :)

iarwain commented:
@hcarty Review flow aside, is there any other topic/concern we need to address before merging or am I clear to proceed?

hcarty commented:
@iarwain No concerns from my side. You are clear to proceed. Thank you again!

iarwain commented:
Great, and thank you for all the work! Merging now.
